src/dawn/Serialization/IIRSerializer.cpp

```diff
@@ -542,16 +588,19 @@ void IIRSerializer::deserializeMetaData(std::shared_ptr<iir::StencilInstantiatio
       auto field = makeField(protoField);
       sirStencilCall->Args.push_back(field);
     }
-    auto stmt = std::make_shared<iir::StencilCallDeclStmt>(sirStencilCall,
-                                                      makeLocation(call.stencil_call_decl_stmt()));
-    stmt->setID(call.stencil_call_decl_stmt().id());
+
+    auto stmt = declStmtFinder.stencilCallDecl[call.stencil_call_decl_stmt().id()];
     metadata.insertStencilCallStmt(stmt, IDToCall.first);
   }
 
   for(auto FieldnameToBC : protoMetaData.fieldnametoboundarycondition()) {
-    std::shared_ptr<iir::BoundaryConditionDeclStmt> bc =
-        dyn_pointer_cast<iir::BoundaryConditionDeclStmt>(makeStmt((FieldnameToBC.second)));
-    metadata.fieldnameToBoundaryConditionMap_[FieldnameToBC.first] = bc;
+    auto foundDecl = declStmtFinder.boundaryConditionDecl.find(
+        FieldnameToBC.second.boundary_condition_decl_stmt().id());
+
+    metadata.fieldnameToBoundaryConditionMap_[FieldnameToBC.first] =
+        foundDecl != declStmtFinder.boundaryConditionDecl.end()
+            ? foundDecl->second
+            : dyn_pointer_cast<ast::BoundaryConditionDeclStmt>(makeStmt((FieldnameToBC.second)));
```

Comment:
Since it is serializing IIR, here you need `iir::BoundaryConditionDeclStmt`. See comments above.